### PR TITLE
parse cert to get expire time 

### PR DESCRIPTION
**Pull request by quanjielin**
https://github.com/istio/istio/issues/12674

### CODE REVIEW

**Comment by wattli** [comment]
Add some description of the relevant context/issue?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 1]
```diff
@@ -521,6 +523,21 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+	// Cert exipre time by default is createTime + sc.configOptions.SecretTTL.
+	expireTime := t.Add(sc.configOptions.SecretTTL)
+	block, _ := pem.Decode(certChain)
+	if block != nil {
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err == nil {
+			expireTime = cert.NotAfter
+			log.Debugf("Cert expire time is %v", expireTime)
+		} else {
+			log.Debugf("Failed to parse certificate %v, use default cert TTL", err.Error())
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: throw Errorf instead?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 2]
```diff
@@ -521,6 +523,21 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+		}
+	} else {
+		log.Debug("Failed to decode certificate, use default cert TTL")
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: Same here.

Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: Do we need any extra actions if we can't get a valid expTime from a cert? Apparently we can't use this cert for service-to-service communication any more.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 1 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: current logic is CA respect secret TTL passed from nodeagent and uses it as cert TTL, this PR is from ask in https://github.com/istio/istio/issues/12674, the updated logic is
1. nodeagent parse cert to get TTL(to handle case that customer CA don't respect TTL passed to it)
2. fallback to default configured TTL if cert parse fail(for some reason)
so I think using debug log is fine.

File: security/pkg/nodeagent/cache/secretcache.go (hunk 2 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: current logic is CA respect secret TTL passed from nodeagent and uses it as cert TTL; this PR is handle CA that don't respect the TTL passed to it, and we fallback to default configured TTL if cert parse fail(for some reason)

Comment (wattli) [reply security/pkg/nodeagent/cache/secretcache.go]: I think the question is more on cert itself (not TTL). I assume we are receiving a cert without a valid TTL in this case. Then should we still let envoy to use this cert for mTLS communication?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 3]
```diff
@@ -521,6 +526,22 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+	expireTime := t.Add(sc.configOptions.SecretTTL)
+	if !sc.configOptions.SkipValidateCert {
+		block, _ := pem.Decode(certChain)
+		if block == nil {
+			log.Errorf("Failed to decode certificate for %q", resourceName)
+			return nil, errors.New("failed to decode certificate")
+		}
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			log.Errorf("Failed to parse certificate for %q: %v", resourceName, err)
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: log the cert?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 4]
```diff
@@ -521,6 +526,22 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
 		certChain = append(certChain, []byte(c)...)
 	}
 
+	// Cert exipre time by default is createTime + sc.configOptions.SecretTTL.
+	expireTime := t.Add(sc.configOptions.SecretTTL)
+	if !sc.configOptions.SkipValidateCert {
+		block, _ := pem.Decode(certChain)
+		if block == nil {
+			log.Errorf("Failed to decode certificate for %q", resourceName)
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: log the cert and err message?

File: security/pkg/nodeagent/cache/secretcache.go [hunk 5]
```diff
@@ -521,6 +526,22 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
 		certChain = append(certChain, []byte(c)...)
 	}
 
+	// Cert exipre time by default is createTime + sc.configOptions.SecretTTL.
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: Add more comment to explain "by default" behavior and non-default behavior?

File: security/pkg/nodeagent/cache/secretcache.go (hunk 4 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: done

File: security/pkg/nodeagent/cache/secretcache.go (hunk 5 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: done

File: security/pkg/nodeagent/cache/secretcache.go [hunk 6]
```diff
@@ -527,6 +532,24 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+	// Current CAs(Citadel, GoogleCA) respect SecretTTL that passed to it and use it decide TTL of cert it issued.
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: Please remove GoogleCA

File: security/pkg/nodeagent/cache/secretcache.go [hunk 7]
```diff
@@ -527,6 +532,24 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token, resourceName s
...
+	// Some customer CA may override TTL param that's passed to it.
+	expireTime := t.Add(sc.configOptions.SecretTTL)
+	if !sc.configOptions.SkipValidateCert {
+		block, _ := pem.Decode(certChain)
```
Comment (wattli) [inline security/pkg/nodeagent/cache/secretcache.go]: Replace "_" with "err" and print it out in the log?

File: security/pkg/nodeagent/cache/secretcache.go (hunk 6 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: auto-merge, will do in another pr..

File: security/pkg/nodeagent/cache/secretcache.go (hunk 7 above)
Comment (quanjielin) [reply security/pkg/nodeagent/cache/secretcache.go]: the second ouput isn't err.

**Pull request merged**